### PR TITLE
Add model fit for FasterRCNN.

### PR DESCRIPTION
**Pull request by tanzhenyu**
Completing the story of `Model.fit` for FasterRCNN.
This is achieved by overriding `train_step` and `test_step`.
Testing it with Pascal VOC dataset, the performance / quality is similar to CTL.
Currently I'm not adding functionality to upload weights and benchmark test given we don't have a metrics system that produces the same result as pycoco yet.

Some design questions:
1) the compile method is following similar pattern of RetinaNet, i.e., compile(box_loss, cls_loss, ...). I think this is fine, but would like to have a consensus here.
2) the compile method uses `weight_decay` -- not too happy about it, but we cannot ask users to add `kernel_regularizer` for their backbones at this point. I think in TF2.10 we may be able to rely on `optimizers.SGDW` instead, but given weight decay should skip batch norm, and KerasCV should work with tf2.9 (I believe there's a huge optimizer bug fix which didn't get backported to 2.9), this seems to be the only way.

### CODE REVIEW

File: examples/models/object_detection/faster_rcnn/train.py [hunk 1]
```diff
@@ -18,16 +18,38 @@
...
-strategy = tf.distribute.MirroredStrategy()
+# Try to detect an available TPU. If none is present, default to MirroredStrategy
+try:
+    tpu = tf.distribute.cluster_resolver.TPUClusterResolver.connect()
+    strategy = tf.distribute.TPUStrategy(tpu)
+except ValueError:
+    # MirroredStrategy is best for a single machine with one or multiple GPUs
+    strategy = tf.distribute.MirroredStrategy()
+print("Number of accelerators: ", strategy.num_replicas_in_sync)
+
```
Comment (innat) [inline examples/models/object_detection/faster_rcnn/train.py:52]: It would be better to have a `keras.utility` function to automate the device selection.

File: keras_cv/models/object_detection/faster_rcnn.py [hunk 2]
```diff
@@ -346,81 +345,165 @@ def _call_rpn(self, images, training=None):
...
-                anchor_format="yxyx",
-                box_format=self.bounding_box_format,
-                variance=[0.1, 0.1, 0.2, 0.2],
-            )
-            res["rcnn_box_pred"] = rcnn_box_pred
-            res["rcnn_cls_pred"] = rcnn_cls_pred
 
-        return res
+    def train_step(self, data):
+        images = data["images"]
```
Comment (ianstenbit) [inline keras_cv/models/object_detection/faster_rcnn.py:478]: It seems like this is a bit different from the current RetinaNet box API.

IIRC we wanted to move to dictionary inputs due to the awkwardness of having box tensors include a class and confidence?

I don't have a strong opinion other than that it should be consistent between the two models

File: examples/models/object_detection/faster_rcnn/train.py [hunk 3]
```diff
@@ -241,13 +263,13 @@ def pad_fn(examples):
 with strategy.scope():
     # The keras huber loss will sum all losses and divide by BS * N * 4, instead we want divide by BS
     # using NONE reduction as also summing creates issues
-    rpn_reg_loss_fn = tf.keras.losses.Huber(reduction=tf.keras.losses.Reduction.NONE)
+    rpn_reg_loss_fn = tf.keras.losses.Huber(reduction=tf.keras.losses.Reduction.SUM)
```
Comment (ianstenbit) [inline examples/models/object_detection/faster_rcnn/train.py:266]: It seems like these loss_fns are no longer used in this script since we're using the defaults from `compile`?

File: keras_cv/models/object_detection/faster_rcnn.py [hunk 4]
```diff
@@ -346,81 +345,165 @@ def _call_rpn(self, images, training=None):
...
+
+    def test_step(self, data):
+        images = data["images"]
+        gt_boxes = data["gt_boxes"]
+        gt_classes = data["gt_classes"]
+        self.compute_loss(images, gt_boxes, gt_classes, training=False)
+        return self.compute_metrics(images, {}, {}, sample_weight={})
+
+
+def _validate_loss(loss, loss_name):
```
Comment (ianstenbit) [inline keras_cv/models/object_detection/faster_rcnn.py:501]: Should we also validate that, for losses which require a box format, the format matches the format of the rest of the model?

File: keras_cv/models/object_detection/faster_rcnn.py (hunk 2 above)
Comment (LukeWood) [inline keras_cv/models/object_detection/faster_rcnn.py:478]: instead of data being a dictionary, lets make data a tuple of (x, y) like a standard Keras model.

Let's make `y` a dictionary with keras {"boxes", "classes"}.  We can migrate RetinaNet to the same format, alongside COCO metrics, and all other components (KPLs, etc)

**Review by LukeWood — CHANGES_REQUESTED** [review]
Minor comments, I see Ian left a review so to avoid spamming I'll keep the current one brief until more changes are pushed.

File: keras_cv/models/object_detection/faster_rcnn.py [hunk 5]
```diff
@@ -346,81 +345,165 @@ def _call_rpn(self, images, training=None):
...
+        classification_loss=None,
+        rpn_box_loss=None,
+        rpn_classification_loss=None,
+        weight_decay=0.0001,
+        **kwargs,
+    ):
+        # TODO(tanzhenyu): Add metrics support once COCOMap issue is addressed.
+        # https://github.com/keras-team/keras-cv/issues/915
+        if "metrics" in kwargs.keys():
+            raise ValueError("currently metrics support is not supported intentionally")
```
[the remark this answers is not in the document]
Comment (tanzhenyu) [reply keras_cv/models/object_detection/faster_rcnn.py:390]: metrics defined here are for both train-time and eval

File: keras_cv/models/object_detection/faster_rcnn.py [hunk 6]
```diff
@@ -346,81 +345,165 @@ def _call_rpn(self, images, training=None):
...
-        return rcnn_box_pred, rcnn_cls_pred
+        return box_pred, cls_pred
+
+    def compile(
+        self,
+        box_loss=None,
+        classification_loss=None,
+        rpn_box_loss=None,
+        rpn_classification_loss=None,
+        weight_decay=0.0001,
```
[the remark this answers is not in the document]
Comment (tanzhenyu) [reply keras_cv/models/object_detection/faster_rcnn.py:384]: Unfortunately our optimizer doesn't take that as argument. We can use SGDW in tf 2.10, but it's actually quite limited

File: examples/models/object_detection/faster_rcnn/train.py (hunk 3 above)
Comment (tanzhenyu) [reply examples/models/object_detection/faster_rcnn/train.py:266]: Good catch! Removed.

File: keras_cv/models/object_detection/faster_rcnn.py (hunk 4 above)
Comment (tanzhenyu) [reply keras_cv/models/object_detection/faster_rcnn.py:501]: loss is independent to box format, IIUC. So users should be able to change between formats and still use the same loss. However they would have to scale differently

File: keras_cv/models/object_detection/faster_rcnn.py [hunk 7]
```diff
@@ -346,81 +345,165 @@ def _call_rpn(self, images, training=None):
...
+        **kwargs,
+    ):
+        # TODO(tanzhenyu): Add metrics support once COCOMap issue is addressed.
+        # https://github.com/keras-team/keras-cv/issues/915
+        if "metrics" in kwargs.keys():
+            raise ValueError("currently metrics support is not supported intentionally")
+        reduction_type = tf.keras.losses.Reduction.SUM
+        if not box_loss:
+            box_loss = tf.keras.losses.Huber(reduction=reduction_type)
```
[the remark this answers is not in the document]
Comment (tanzhenyu) [reply keras_cv/models/object_detection/faster_rcnn.py:393]: Sounds good. Done. We should also do this for RetinaNet then.

File: examples/models/object_detection/faster_rcnn/train.py [hunk 8]
```diff
@@ -18,16 +18,38 @@
...
+try:
+    tpu = tf.distribute.cluster_resolver.TPUClusterResolver.connect()
+    strategy = tf.distribute.TPUStrategy(tpu)
+except ValueError:
+    # MirroredStrategy is best for a single machine with one or multiple GPUs
+    strategy = tf.distribute.MirroredStrategy()
+print("Number of accelerators: ", strategy.num_replicas_in_sync)
+
 local_batch = 4
 global_batch = local_batch * strategy.num_replicas_in_sync
```
[the remark this answers is not in the document]
Comment (tanzhenyu) [reply examples/models/object_detection/faster_rcnn/train.py:54]: global_batch is still used in data pipeline, and also used to scale our optimizer lr

File: keras_cv/models/object_detection/faster_rcnn.py (hunk 2 above)
Comment (tanzhenyu) [reply keras_cv/models/object_detection/faster_rcnn.py:478]: Good point. Done.

File: examples/models/object_detection/faster_rcnn/train.py (hunk 1 above)
Comment (tanzhenyu) [reply examples/models/object_detection/faster_rcnn/train.py:52]: To me this sounds more like a `tf.distribute.auto_select_strategy` :-)

Comment (tanzhenyu) [reply examples/models/object_detection/faster_rcnn/train.py:52]: Having a trainer object that takes care of runtime args is definitely not a bad idea

File: keras_cv/models/object_detection/faster_rcnn.py (hunk 4 above)
Comment (ianstenbit) [reply keras_cv/models/object_detection/faster_rcnn.py]: Is IOU loss not relevant for this model? (IIRC that requires a box format?)

File: examples/models/object_detection/faster_rcnn/train.py (hunk 1 above)
Comment (LukeWood) [reply examples/models/object_detection/faster_rcnn/train.py:52]:
```
trainer = Trainer(devices=8, accelerator="gpu")
```

isn't this just `MirroredStrategy(devices=8)`

File: keras_cv/models/object_detection/faster_rcnn.py (hunk 4 above)
Comment (tanzhenyu) [reply keras_cv/models/object_detection/faster_rcnn.py]: There seem to be inconsistent losses: 1) iou loss that require a format; 2) smoothl1 loss that doesn't require a format.
In the long run, we should probably by fixing it so that it's format independent. Why? Because there are some weird format (and actually mostly used compared to other formats) such as `tx` `ty` `tw` `th` which we wouldn't be able to scale up by supporting all of them, and clearly we don't want to block users if they want to convert their format to a new one before passing to loss function.

File: examples/models/object_detection/faster_rcnn/train.py (hunk 1 above)
Comment (innat) [reply examples/models/object_detection/faster_rcnn/train.py:52]: @LukeWood you're right. Now, the following code is widely used. 

```python
try:
    tpu = tf.distribute.cluster_resolver.TPUClusterResolver.connect()
    strategy = tf.distribute.TPUStrategy(tpu)
except ValueError:
    # MirroredStrategy is best for a single machine with one or multiple GPUs
    strategy = tf.distribute.MirroredStrategy()
print("Number of accelerators: ", strategy.num_replicas_in_sync)
```

That's why I was thinking it might be good to have a simple utility function to automate it, simple and clean.

i.e. `keras.utils.set_strategy(devices=2, accelerator="gpu")`

File: keras_cv/models/object_detection/faster_rcnn.py [hunk 9]
```diff
@@ -346,81 +346,169 @@ def _call_rpn(self, images, training=None):
...
+        # TODO(tanzhenyu): Add metrics support once COCOMap issue is addressed.
+        # https://github.com/keras-team/keras-cv/issues/915
+        if "metrics" in kwargs.keys():
+            raise ValueError("currently metrics support is not supported intentionally")
+        box_loss = _validate_and_get_loss(box_loss, "box_loss")
+        classification_loss = _validate_and_get_loss(
+            classification_loss, "classification_loss"
+        )
+        rpn_box_loss = _validate_and_get_loss(rpn_box_loss, "rpn_box_loss")
+        if rpn_classification_loss == "BinaryCrossentropy":
```
Comment (LukeWood) [inline keras_cv/models/object_detection/faster_rcnn.py:398]: can we move this logic into _validate_and_get_loss?

**Review by LukeWood — CHANGES_REQUESTED** [review]

File: keras_cv/models/object_detection/faster_rcnn.py [hunk 10]
```diff
@@ -346,81 +346,169 @@ def _call_rpn(self, images, training=None):
...
                 variance=[0.1, 0.1, 0.2, 0.2],
             )
-        return rcnn_box_pred, rcnn_cls_pred
+        return box_pred, cls_pred
+
+    # TODO(tanzhenyu): Support compile with metrics.
+    def compile(
+        self,
+        box_loss="Huber",
+        classification_loss="SparseCategoricalCrossentropy",
```
[the remark this answers is not in the document]
Comment (tanzhenyu) [reply keras_cv/models/object_detection/faster_rcnn.py:383]: This does not support lower cased losses yet

File: keras_cv/models/object_detection/faster_rcnn.py [hunk 11]
```diff
@@ -346,81 +346,169 @@ def _call_rpn(self, images, training=None):
...
                 box_format=self.bounding_box_format,
                 variance=[0.1, 0.1, 0.2, 0.2],
             )
-        return rcnn_box_pred, rcnn_cls_pred
+        return box_pred, cls_pred
+
+    # TODO(tanzhenyu): Support compile with metrics.
+    def compile(
+        self,
+        box_loss="Huber",
```
[the remark this answers is not in the document]
Comment (tanzhenyu) [reply keras_cv/models/object_detection/faster_rcnn.py]: This seems reasonable to me. But `loss='default'` is definitely more weird, changing it to require users to specifically pass each loss.

**Review by LukeWood — APPROVED** [review]

**Review by ianstenbit — APPROVED** [review]

File: examples/models/object_detection/faster_rcnn/train.py [hunk 12]
```diff
@@ -18,16 +18,38 @@
 Last modified: 2022/09/27
 Description: Use KerasCV to train a RetinaNet on Pascal VOC 2007.
```
Comment (qlzh727) [inline examples/models/object_detection/faster_rcnn/train.py:19]: File name wise, since this is train wih pascal voc, how about change it to train_pascal_voc.py?

File: examples/models/object_detection/faster_rcnn/train.py [hunk 13]
```diff
@@ -18,16 +18,38 @@
...
 
 import tensorflow as tf
 import tensorflow_datasets as tfds
+from absl import flags
 
 import keras_cv
 
+flags.DEFINE_string(
+    "weights_path",
+    "weights_{epoch:02d}.h5",
```
[the remark this answers is not in the document]
Comment (qlzh727) [reply examples/models/object_detection/faster_rcnn/train.py:31]: Same here, usually it just need to either keep the last epoch weights, or best epoch weights. Even you have the weights for every epoch, you probably still like to the last or the best epoch weights.

File: examples/models/object_detection/faster_rcnn/train.py [hunk 14]
```diff
@@ -267,167 +269,17 @@ def pad_fn(examples):
 weight_decay = 0.0001
```
Comment (qlzh727) [inline examples/models/object_detection/faster_rcnn/train.py:269]: This one is not used anymore here.

File: examples/models/object_detection/faster_rcnn/train.py [hunk 15]
```diff
@@ -267,167 +269,17 @@ def pad_fn(examples):
...
 step = 0
```
Comment (qlzh727) [inline examples/models/object_detection/faster_rcnn/train.py:270]: Same for this one.

File: examples/models/object_detection/faster_rcnn/train.py [hunk 16]
```diff
@@ -239,22 +257,6 @@ def pad_fn(examples):
...
-        reduction=tf.keras.losses.Reduction.NONE
-    )
-
-    rpn_reg_metric = tf.keras.metrics.Mean()
-    rpn_cls_metric = tf.keras.metrics.Mean()
-    rcnn_reg_metric = tf.keras.metrics.Mean()
-    rcnn_cls_metric = tf.keras.metrics.Mean()
-
     lr_decay = tf.keras.optimizers.schedules.PiecewiseConstantDecay(
         boundaries=[12000 * 16 / global_batch, 16000 * 16 / global_batch],
```
Comment (qlzh727) [inline examples/models/object_detection/faster_rcnn/train.py:261]: i think the boundary here really depends on the size of the dataset (since the ds is not repeat and there isn't a steps per epoch in fit()). We might want to change 12000 and 16000 to something more meaningful (I assume there are 1000 steps per dataset, and 12000 refers to 12 epochs.)

File: keras_cv/models/object_detection/faster_rcnn.py [hunk 17]
```diff
@@ -346,81 +346,176 @@ def _call_rpn(self, images, training=None):
...
+    def compile(
+        self,
+        box_loss=None,
+        classification_loss=None,
+        rpn_box_loss=None,
+        rpn_classification_loss=None,
+        weight_decay=0.0001,
+        loss=None,
+        **kwargs,
+    ):
```
Comment (qlzh727) [inline keras_cv/models/object_detection/faster_rcnn.py:389]: I think we need some docstring for this method.

File: keras_cv/models/object_detection/faster_rcnn.py [hunk 18]
```diff
@@ -346,81 +346,176 @@ def _call_rpn(self, images, training=None):
...
+        # TODO(tanzhenyu): Add metrics support once COCOMap issue is addressed.
+        # https://github.com/keras-team/keras-cv/issues/915
+        if "metrics" in kwargs.keys():
+            raise ValueError("currently metrics support is not supported intentionally")
+        if loss is not None:
+            raise ValueError(
+                "`FasterRCNN` does not accept a `loss` to `compile()`. "
+                "Instead, please pass `box_loss` and `classification_loss`. "
```
Comment (qlzh727) [inline keras_cv/models/object_detection/faster_rcnn.py:397]: how about ` rpn_box_loss` and `rpn_classification_loss`?

File: keras_cv/models/object_detection/faster_rcnn.py [hunk 19]
```diff
@@ -346,81 +346,176 @@ def _call_rpn(self, images, training=None):
...
+
+def _validate_and_get_loss(loss, loss_name):
+    if isinstance(loss, str):
+        loss = tf.keras.losses.get(loss)
+    if loss is None or not isinstance(loss, tf.keras.losses.Loss):
+        raise ValueError(
+            f"FasterRCNN only accepts `tf.keras.losses.Loss` for {loss_name}, got {loss}"
+        )
+    if loss.reduction != tf.keras.losses.Reduction.SUM:
+        logging.info(
```
Comment (qlzh727) [inline keras_cv/models/object_detection/faster_rcnn.py:517]: Should this be a info or a warning?

**Review by qlzh727 — DISMISSED** [review]

File: keras_cv/models/object_detection/faster_rcnn.py (hunk 2 above)
Comment (LukeWood) [reply keras_cv/models/object_detection/faster_rcnn.py]: Can we drop `'gt_'`?

File: keras_cv/models/object_detection/faster_rcnn.py [hunk 20]
```diff
@@ -346,81 +346,176 @@ def _call_rpn(self, images, training=None):
...
-            )
-            res["rcnn_box_pred"] = rcnn_box_pred
-            res["rcnn_cls_pred"] = rcnn_cls_pred
 
-        return res
+    def train_step(self, data):
+        images, y, sample_weight = tf.keras.utils.unpack_x_y_sample_weight(data)
+        if sample_weight is not None:
+            raise ValueError("`sample_weight` is currently not supported.")
+        gt_boxes = y["gt_boxes"]
```
Comment (LukeWood) [inline keras_cv/models/object_detection/faster_rcnn.py:485]: lets drop gt_ in the dictionary keys.

Its already implied as they belong to `y`.

Comment (tanzhenyu) [reply keras_cv/models/object_detection/faster_rcnn.py:485]: Oops missed this one before I merge.This sounds good to me. Though if we add `masks` in the future, `gt_masks` seems more
